### PR TITLE
Update error message when catalog entry is invalid

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description
Fix https://github.com/kedro-org/kedro/issues/3910

## Development notes

Add an error message if the catalog entry is not a dictionary - eg
```yaml
invalid_entry: whatever
```
Also, when the catalog entry is a dict but `type:` is missing.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/io/data_catalog.py [hunk 1]
```diff
@@ -288,6 +288,12 @@ class to be loaded is specified with the key ``type`` and their
         user_default = {}
 
         for ds_name, ds_config in catalog.items():
+            if not isinstance(ds_config, dict):
```
Comment (merelcht) [inline kedro/io/data_catalog.py:291]: In https://github.com/kedro-org/kedro/issues/3555 there was a bit of discussion on what to do with values that are dict type, but not meant as dataset. Is it worth updating the message for that case as well to mention interpolation?

**Review by merelcht — COMMENTED** [review]
I left one question. + don't forget to add this to the release notes ✍️

File: kedro/io/data_catalog.py (hunk 1 above)
Comment (ankatiyar) [reply kedro/io/data_catalog.py:291]: Yep, I've added a hint to the error that comes from `AbstractDataset.from_config()` as well!

**Review by merelcht — APPROVED** [review]
LGTM 👍

**Review by DimedS — COMMENTED** [review]
Thank you, @ankatiyar ! The PR looks great. I have one question: Is it okay that we receive a different error if we put just `some_value` into the catalog without a `:` after it?

**Comment by ankatiyar** [comment]
> Thank you, @ankatiyar ! The PR looks great. I have one question: Is it okay that we receive a different error if we put just `some_value` into the catalog without a `:` after it?

Some offline discussion with @DimedS on this. If there is a catalog entry which is just one word like this-
```yaml
some_value
```
This is wrong YAML syntax and it'll error out much earlier when `OmegaConfigLoader` tries to load the catalog. -
```
 File "/Users/ankita_katiyar/kedro/kedro/kedro/config/omegaconf_config.py", line 319, in load_and_merge_dir_config
    raise ParserError(
yaml.parser.ParserError: Invalid YAML or JSON file /Users/ankita_katiyar/kedro_projects/demo/conf/base/catalog.yml, unable to read line 73, position 0.
```
This is not Kedro specific, it simply is just not allowed in YAML afaik. So to me, they seem like two different issues. This PR is to address when a catalog entry is correct syntactically in YAML but "wrong" as per Kedro rules (eg. not a valid dataset) in `catalog.yml`

**Review by DimedS — APPROVED** [review]
I agree. It seems it will be complicated to handle this properly on the Kedro side. Additionally, the error is meaningful as it points to the specific line number in `catalog.yml` where the parsing issue occurs, unlike the other error mentioned in the ticket.

**Pull request merged**